### ISSUE TITLE
IPv6 addresses don't work unless port is explicitly specified

### ISSUE THREAD

**Issue by Splat** [capistrano/sshkit#403]
I'm sure this is an issue on my end that I'm fundamentally misunderstanding. I can SSH into a server with an address like so:
```
2001:aaa:bbb:cc:dddd:eeee:ffff:gggg
```

There is no issues until I try to leverage sshkit to do the same. 
```
on ["[2001:aaa:bbb:cc:dddd:eeee:ffff:gggg]"], in: :parallel do |host|
```

What I end up seeing is:
```
Exception on host 2001:aaa:bbb:cc:dddd:eeee:ffff caught: getaddrinfo: nodename nor servname provided, or not known
```
Note the trailing 16bits seems to be truncated off? Is there a different way I should be formatting that IPv6 address string?

**Comment by leehambley** [comment]
It's entirely possible that it's truncated by the logger, not the underlying system.

All our SSH code is deferred to http://net-ssh.github.io/net-ssh/ which does the heavy lifting.

Could you please try a minimal reproduction case, and see if this issue should rather be filed upstream?

``` ruby 
require 'net/ssh'

Net::SSH.start('host', 'user', :password => "password") do |ssh|
  # capture all stderr and stdout output from a remote process
  output = ssh.exec!("hostname")
  puts output

  # capture only stdout matching a particular pattern
  stdout = ""
  ssh.exec!("ls -l /home/jamis") do |channel, stream, data|
    stdout << data if stream == :stdout
  end
  puts stdout

  # run multiple processes in parallel to completion
  ssh.exec "sed ..."
  ssh.exec "awk ..."
  ssh.exec "rm -rf ..."
  ssh.loop

  # open a new channel and configure a minimal set of callbacks, then run
  # the event loop until the channel finishes (closes)
  channel = ssh.open_channel do |ch|
    ch.exec "/usr/local/bin/ruby /path/to/file.rb" do |ch, success|
      raise "could not execute command" unless success

      # "on_data" is called when the process writes something to stdout
      ch.on_data do |c, data|
        $stdout.print data
      end

      # "on_extended_data" is called when the process writes something to stderr
      ch.on_extended_data do |c, type, data|
        $stderr.print data
      end

      ch.on_close { puts "done!" }
    end
  end

  channel.wait

  # forward connections on local port 1234 to port 80 of www.capify.org
  ssh.forward.local(1234, "www.capify.org", 80)
  ssh.loop { true }
end
```

**Comment by Splat** [comment]
Definitely will do. I'll report findings and close if the case.

**Comment by mattbrictson** [comment]
There does seem to be a legitimate bug in how `SSHKit::Host` parses IPv6 address. Based on my testing, the port number must be explicitly specified in the host string. Otherwise it interprets the last `:XXXX` part of the address to be the port number, which is incorrect.

This is a workaround:

```ruby
# Note that port number must be specified outside the [] brackets
on "[2001:db8:85a3:8d3:1319:8a2e:370:7348]:22" do |host|
  # ...
end
```

This illustrates the bug:

```ruby
# With port number it works
SSHKit::Host.new '[2001:db8:85a3:8d3:1319:8a2e:370:7348]:22'
# => #<SSHKit::Host:0x007fc1231e46c8 @keys=[], @local=false, @user=nil, @hostname="2001:db8:85a3:8d3:1319:8a2e:370:7348", @port=22>

# Without port number, it parses incorrectly
SSHKit::Host.new '[2001:db8:85a3:8d3:1319:8a2e:370:7348]'
# => #<SSHKit::Host:0x007fc12318d7d8 @keys=[], @local=false, @user=nil, @hostname="2001:db8:85a3:8d3:1319:8a2e:370", @port=7348>
```

**Comment by mattbrictson** [comment]
This should be a fairly straightforward fix, if someone wants to volunteer to submit a PR! 🙏

**Comment by Splat** [comment]
I'd be happy to take care of it @mattbrictson

**Comment by mattbrictson** [comment]
Much appreciated! Let me know if you have any questions, and please refer to our CONTRIBUTING.md doc for general guidelines.

**Comment by TheRusskiy** [comment]
@mattbrictson was this fixed somewhere?

**Comment by will-in-wi** [comment]
Looks like not:
```
irb(main):001:0> require 'sshkit'
=> true
irb(main):002:0> SSHKit::Host.new '[2001:db8:85a3:8d3:1319:8a2e:370:7348]:22'
=> #<SSHKit::Host:0x00007f8f4eae9d88 @keys=[], @local=false, @user=nil, @hostname="2001:db8:85a3:8d3:1319:8a2e:370:7348", @port=22>
irb(main):003:0> SSHKit::Host.new '[2001:db8:85a3:8d3:1319:8a2e:370:7348]'
=> #<SSHKit::Host:0x00007f8f4ead3290 @keys=[], @local=false, @user=nil, @hostname="2001:db8:85a3:8d3:1319:8a2e:370", @port=7348>
```

I'll see if it is easy to whip up a patch.
